### PR TITLE
feat(bucket): unique field option

### PR DESCRIPTION
**Pull request by htuna07**

### CODE REVIEW

**Comment by thesayyn** [comment]
Can we hold this PR until we land bucket rules? @htuna07.

There will be enormous conflicts between these PRs.

File: stacks/api/bucket/services/bucket.service.ts [hunk 1]
```diff
@@ -23,6 +23,17 @@ export class BucketService extends BaseCollection<Bucket>("buckets") {
     this.schemaChangeEmitter.next(undefined);
   }
 
+  async insertOne(bucket: Bucket) {
+    const insertedBucket = await this.buckets.insertOne(bucket).then(t => t.ops[0]);
+    const bucketCollection = await this.db.createCollection(`bucket_${insertedBucket._id}`);
```
Comment (thesayyn) [inline stacks/api/bucket/services/bucket.service.ts:28]: You can use getBucketDataCollection  from "@spica-server/bucket/services"

File: stacks/api/bucket/services/bucket.service.ts [hunk 2]
```diff
@@ -23,6 +23,17 @@ export class BucketService extends BaseCollection<Bucket>("buckets") {
     this.schemaChangeEmitter.next(undefined);
   }
 
+  async insertOne(bucket: Bucket) {
+    const insertedBucket = await this.buckets.insertOne(bucket).then(t => t.ops[0]);
```
Comment (thesayyn) [inline stacks/api/bucket/services/bucket.service.ts:27]: Also, you could use  super.insertOne().

File: stacks/api/bucket/src/crud.ts [hunk 3]
```diff
@@ -200,7 +204,16 @@ export async function insertDocument(
     throw new ForbiddenException("ACL rules has rejected this operation.");
   }
 
-  return collection.insertOne(document);
+  return collection.insertOne(document).catch(error => {
```
Comment (thesayyn) [inline stacks/api/bucket/src/crud.ts:207]: You could return a message like `value of the property .title should unique across all documents.`

File: stacks/api/bucket/src/crud.ts [hunk 4]
```diff
@@ -232,9 +245,20 @@ export async function replaceDocument(
...
-  return collection.findOneAndReplace({_id: documentId}, document, {
-    returnOriginal: options.returnOriginal
-  });
+  return collection
+    .findOneAndReplace({_id: documentId}, document, {
+      returnOriginal: options.returnOriginal
+    })
+    .catch(error => {
+      if (error.code === 11000) {
+        throw new BadRequestException(
```
Comment (thesayyn) [inline stacks/api/bucket/src/crud.ts:254]: Same as above.

File: stacks/api/bucket/src/crud.ts [hunk 5]
```diff
@@ -266,9 +290,20 @@ export async function patchDocument(
...
   const updateQuery = getUpdateQueryForPatch(patch);
 
-  return collection.findOneAndUpdate({_id: document._id}, updateQuery, {
-    returnOriginal: options.returnOriginal
-  });
+  return collection
+    .findOneAndUpdate({_id: document._id}, updateQuery, {
+      returnOriginal: options.returnOriginal
+    })
+    .catch(error => {
```
Comment (thesayyn) [inline stacks/api/bucket/src/crud.ts:297]: same as above

**Review by thesayyn — CHANGES_REQUESTED** [review]

**Comment by thesayyn** [comment]
you should wait for the https://github.com/spica-engine/spica/pull/521 first.

**Comment by thesayyn** [comment]
ping @htuna07

**Review by akhnos — APPROVED** [review]

**Review by thesayyn — APPROVED** [review]
lgtm

**Pull request merged**